File: my-folder/0238-product-of-array-except-self/solution.py
class Solution:
    def productExceptSelf(self, nums: List[int]) -> List[int]:
        pro = 1
        res = [1] * len(nums)

        for i in range(len(nums)):
            res[i] = pro
            pro *= nums[i]

        pro = 1

        for i in range(len(nums) - 1, -1, -1):
            res[i] *= pro
            pro *= nums[i]
        
        return res

File: my-folder/0238-product-of-array-except-self/solution.java
class Solution {
    public int[] productExceptSelf(int[] nums) {
        int pre = 1;
        int suf = 1;
        int[] prefix = new int[nums.length];
        int[] suffix = new int[nums.length];
        for (int i = 0; i < prefix.length; i++) {
            prefix[i] = pre;
            pre *= nums[i];
        }
        for (int i = nums.length - 1; i >= 0; i--) {
            suffix[i] = suf;
            suf *= nums[i];
        }
        int[] res = new int[nums.length];
        for (int i = 0; i < res.length; i++) {
            res[i] = prefix[i] * suffix[i];
        }
        return res;
    }
}
